### ISSUE TITLE
work incorrectly,while gl_SubgroupSize != 32

### ISSUE THREAD

**Issue by MouseChannel** [jaesung-cs/vulkan_radix_sort#7]

**Comment by jaesung-cs** [comment]
Sorry, I don't have a GPU with subgroup size != 32, and don't have a plan to support for the case in near future.

**Comment by Garmaz** [comment]
> Sorry, I don't have a GPU with subgroup size != 32, and don't have a plan to support for the case in near future.

Could you give some suggestions on what I should change in the code to support, say subgroup size =64? I'm having trouble debugging my program, the shader code is hard to understand :(

**Comment by jaesung-cs** [comment]
Could you first try change SUBGROUP_SIZE define from 32 to 64?

2025년 1월 2일 (목) 오후 1:49, Garmaz ***@***.***>님이 작성:

**Comment by jaesung-cs** [comment]
Could you first try change SUBGROUP_SIZE define from 32 to 64?

**Comment by Garmaz** [comment]
> Could you first try change SUBGROUP_SIZE define from 32 to 64?
Sure I have tried and still the output is not as expected, I kept on getting key == value in the output. 
Actually, I have also changed the workgroup size to 256 and partition division to 16, otherwise I would get the error of VK_ERROR_DEVICE_LOST

**Comment by jaesung-cs** [comment]
Thanks for the clue.
Could you confirm if key-only sort works correctly and key-value sort?
I will investigate this issue further this week.

**Comment by jaesung-cs** [comment]
Could you try `feature/bench_apple` branch?
https://github.com/jaesung-cs/vulkan_radix_sort/tree/feature/bench_apple
Last year I've renewed benchmark code and fixed key-value sort (check out this commit: https://github.com/jaesung-cs/vulkan_radix_sort/commit/40dcfb4a06c3ac97c54a128136869680882edd83), but forgot to merge.
I just tested with NVIDIA GeForce RTX 4090, subgroupSize=32, on Windows
```ps1
> cmake . -B build
> cmake --build build --config Release -j
> .\build\Release\bench.exe 10000000 vulkan
vk_radix_sort benchmark
================ sort ================
total time: 0.760544ms (13.1485 GItems/s)
================ sort key value ================
total time: 0.970752ms (10.3013 GItems/s)
================ sort key value speed ================
[0] total time: 0.9728ms (10.2796 GItems/s)
[1] total time: 0.970752ms (10.3013 GItems/s)
[2] total time: 0.973664ms (10.2705 GItems/s)
[3] total time: 0.971456ms (10.2938 GItems/s)
[4] total time: 0.971776ms (10.2904 GItems/s)
[5] total time: 0.970592ms (10.303 GItems/s
...
```
and could you share your device info with `vulkaninfo` that you think might be useful, if you don't mind? for example:
```
	deviceName        = NVIDIA GeForce RTX 4090
	maxComputeSharedMemorySize                      = 49152
	maxComputeWorkGroupSize: count = 3
		1024
		1024
		64
	subgroupSize                      = 32
```

**Comment by Garmaz** [comment]
> Could you try `feature/bench_apple` branch? https://github.com/jaesung-cs/vulkan_radix_sort/tree/feature/bench_apple Last year I've renewed benchmark code and fixed key-value sort (check out this commit: [40dcfb4](https://github.com/jaesung-cs/vulkan_radix_sort/commit/40dcfb4a06c3ac97c54a128136869680882edd83)), but forgot to merge. I just tested with NVIDIA GeForce RTX 4090, subgroupSize=32, on Windows
> 
> ```powershell
> > cmake . -B build
> > cmake --build build --config Release -j
> > .\build\Release\bench.exe 10000000 vulkan
> vk_radix_sort benchmark
> ================ sort ================
> total time: 0.760544ms (13.1485 GItems/s)
> ================ sort key value ================
> total time: 0.970752ms (10.3013 GItems/s)
> ================ sort key value speed ================
> [0] total time: 0.9728ms (10.2796 GItems/s)
> [1] total time: 0.970752ms (10.3013 GItems/s)
> [2] total time: 0.973664ms (10.2705 GItems/s)
> [3] total time: 0.971456ms (10.2938 GItems/s)
> [4] total time: 0.971776ms (10.2904 GItems/s)
> [5] total time: 0.970592ms (10.303 GItems/s
> ...
> ```
> 
> and could you share your device info with `vulkaninfo` that you think might be useful, if you don't mind? for example:
> 
> ```
> 	deviceName        = NVIDIA GeForce RTX 4090
> 	maxComputeSharedMemorySize                      = 49152
> 	maxComputeWorkGroupSize: count = 3
> 		1024
> 		1024
> 		64
> 	subgroupSize                      = 32
> ```

Sry for this late reply, was busy. My device info is as follows:
maxComputeSharedMemorySize = {uint32_t} 32768
maxComputeWorkGroupSize = {uint32_t[3]} [3]
    [0] = {uint32_t} 512
    [1] = {uint32_t} 512
    [2] = {uint32_t} 512
subgroupSize = {uint32_t} 64

I will try the bench later on.

 BTW, I think for subgroup size == 64, the GetExclusiveSubgroupMask function in downsweep will have undefined behaviors? Also, is there a latent constraint that RADIX / SUBGROUP_SIZE == PARTITION_DIVISION?

**Comment by jaesung-cs** [comment]
> My device info is as follows:

- Thanks for the information. Shader codes that require 16384 + some more bytes won't exceed your shared memory limit. Same for workgroup size.

> the GetExclusiveSubgroupMask function in downsweep will have undefined behaviors?

- You are right. I read GLSL spec and found:
  > The shift operators (<<) and (>>). ... The result is **undefined** if **the right operand is negative**, or **greater than or
equal to the number of bits** in the left expression’s base type.

  So, let's fix the function to work for id = 0..63.

> Also, is there a latent constraint that RADIX / SUBGROUP_SIZE == PARTITION_DIVISION?

- No, there's no such constraint.

And I found a potential reason for error:
https://github.com/jaesung-cs/vulkan_radix_sort/blob/master/src/shader/downsweep.comp#L163-L170
where `RADIX=256`, `gl_NumSubgroups=8`, `gl_SubgroupSize=64` in your device, thus resulting in `intermediateSize1=0`.

Note that line 138-182 is for calculating exclusive prefix sum of array length 4096 for `SubgroupSize=32`, and length 2048 for `SubgroupSize=64`.

When I was writing the code, I assumed `SubgroupSize=32` case, which requires two stages of reduction by 32, storing exclusive sums in array of length 4096 and sums in array of length 4096/32=128 in the first stage and the same for length 128 and 128/32=4. Then, calculate exclusive sums of array of length 4.

In `SubgroupSize=64` case, `intermediateSize1` is evaluated zero, and the last stage doesn't happen. The exclusive sum of only one element must be zero, but in current code, the value remains still.

A quick fix would be `uint intermediateSize1 = max(..., 1);`

In summary,
- try fix `GetExclusiveSubgroupMask` to work for id = 0..63.
- try `uint intermediateSize1 = max(..., 1);`

**Comment by jaesung-cs** [comment]
here's my fix: https://github.com/jaesung-cs/vulkan_radix_sort/pull/9
could you please test?
